firebase-sessions/src/test/kotlin/com/google/firebase/sessions/SessionInitiatorTest.kt

```diff
@@ -0,0 +1,140 @@
+
+    assertThat(sessionStartCounter.count).isEqualTo(1)
+  }
+
+  @Test
+  fun appForegrounded_largeInterval_initiatesSession() {
+    val fakeTime = FakeTime()
+    val sessionStartCounter = SessionStartCounter()
+
+    val sessionInitiator =
+      SessionInitiator(fakeTime::currentTimeMs, sessionStartCounter::initiateSessionStart)
+
+    // First session on cold start
+    assertThat(sessionStartCounter.count).isEqualTo(1)
+
+    // Enough tome to initiate a new session, and then foreground
+    fakeTime.addTimeMs(LARGE_INTERVAL_MS)
+    sessionInitiator.appForegrounded()
+
+    // Another session initiated
+    assertThat(sessionStartCounter.count).isEqualTo(2)
+  }
+
+  @Test
+  fun appForegrounded_smallInterval_doesNotInitiatesSession() {
+    val fakeTime = FakeTime()
+    val sessionStartCounter = SessionStartCounter()
+
+    val sessionInitiator =
+      SessionInitiator(fakeTime::currentTimeMs, sessionStartCounter::initiateSessionStart)
+
+    // First session on cold start
+    assertThat(sessionStartCounter.count).isEqualTo(1)
+
+    // Not enough time to initiate a new session, and then foreground
+    fakeTime.addTimeMs(SMALL_INTERVAL_MS)
+    sessionInitiator.appForegrounded()
+
+    // No new session
+    assertThat(sessionStartCounter.count).isEqualTo(1)
+  }
+
+  @Test
+  fun appForegrounded_background_foreground_largeIntervals_initiatesSessions() {
+    val fakeTime = FakeTime()
+    val sessionStartCounter = SessionStartCounter()
+
+    val sessionInitiator =
+      SessionInitiator(fakeTime::currentTimeMs, sessionStartCounter::initiateSessionStart)
+
+    assertThat(sessionStartCounter.count).isEqualTo(1)
+
+    fakeTime.addTimeMs(LARGE_INTERVAL_MS)
+    sessionInitiator.appForegrounded()
+
+    assertThat(sessionStartCounter.count).isEqualTo(2)
+
+    sessionInitiator.appBackgrounded()
+    fakeTime.addTimeMs(LARGE_INTERVAL_MS)
+    sessionInitiator.appForegrounded()
+
+    assertThat(sessionStartCounter.count).isEqualTo(3)
+  }
+
+  @Test
+  fun appForegrounded_background_foreground_smallIntervals_doesNotInitiateNewSessions() {
+    val fakeTime = FakeTime()
+    val sessionStartCounter = SessionStartCounter()
+
+    val sessionInitiator =
+      SessionInitiator(fakeTime::currentTimeMs, sessionStartCounter::initiateSessionStart)
+
+    // First session on cold start
+    assertThat(sessionStartCounter.count).isEqualTo(1)
+
+    fakeTime.addTimeMs(SMALL_INTERVAL_MS)
+    sessionInitiator.appForegrounded()
+
+    assertThat(sessionStartCounter.count).isEqualTo(1)
+
+    sessionInitiator.appBackgrounded()
+    fakeTime.addTimeMs(SMALL_INTERVAL_MS)
+    sessionInitiator.appForegrounded()
+
+    assertThat(sessionStartCounter.count).isEqualTo(1)
+
+    assertThat(sessionStartCounter.count).isEqualTo(1)
+  }
+
+  companion object {
+    private const val SMALL_INTERVAL_MS = 3 * 1000L // not enough time to initiate a new session
+    private const val LARGE_INTERVAL_MS = 90 * 60 * 1000L // enough to initiate another session
```

Comment:
Similar to earlier - We should consider using Duration APIs.

Reply:
Done in #4733.

Reply:
Thank you!